From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alan Shearer <sakison@gmail.com>
Date: Tue, 6 Oct 2026 13:40:00 -0500
Subject: [PATCH] Power: preferences window for battery power saving

Add a Power preferences app with two checkboxes. It writes
/boot/system/settings/power_daemon and asks power_daemon to reload.
It does not change the scheduler mode itself.

Power saving on battery is off unless the user turns it on.
Keep power saving on AC stays in power saving after the adapter is
plugged back in.

---
 src/preferences/Jamfile               |   1 +
 src/preferences/power/Jamfile         |  16 +++
 src/preferences/power/Power.rdef      |  14 +++
 src/preferences/power/PowerApp.cpp    |  48 ++++++++
 src/preferences/power/PowerWindow.cpp | 160 ++++++++++++++++++++++++++
 src/preferences/power/PowerWindow.h   |  36 ++++++
 6 files changed, 275 insertions(+)
 create mode 100644 src/preferences/power/Jamfile
 create mode 100644 src/preferences/power/Power.rdef
 create mode 100644 src/preferences/power/PowerApp.cpp
 create mode 100644 src/preferences/power/PowerWindow.cpp
 create mode 100644 src/preferences/power/PowerWindow.h

diff --git a/src/preferences/Jamfile b/src/preferences/Jamfile
index 1111111..2222222 100644
--- a/src/preferences/Jamfile
+++ b/src/preferences/Jamfile
@@ -1,3 +1,4 @@
+SubInclude HAIKU_TOP src preferences power ;
 SubInclude HAIKU_TOP src preferences repositories ;
 SubInclude HAIKU_TOP src preferences screen ;
 SubInclude HAIKU_TOP src preferences screensaver ;
diff --git a/src/preferences/power/Jamfile b/src/preferences/power/Jamfile
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/src/preferences/power/Jamfile
@@ -0,0 +1,16 @@
+SubDir HAIKU_TOP src preferences power ;
+
+Application Power :
+	PowerApp.cpp
+	PowerWindow.cpp
+	: be translation [ TargetLibsupc++ ] localestub
+	: Power.rdef
+	;
+
+DoCatalogs Power :
+	x-vnd.Haiku-Power
+	:
+	PowerApp.cpp
+	PowerWindow.cpp
+	;
+
diff --git a/src/preferences/power/Power.rdef b/src/preferences/power/Power.rdef
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/src/preferences/power/Power.rdef
@@ -0,0 +1,14 @@
+resource app_signature "application/x-vnd.Haiku-Power";
+
+resource app_name_catalog_entry "x-vnd.Haiku-Power:System name:Power";
+
+resource app_version {
+	major  = 1,
+	middle = 0,
+	minor  = 0,
+	variety = 2,
+	internal = 0,
+	short_info = "Power",
+	long_info  = "Power"
+};
+
+resource app_flags B_SINGLE_LAUNCH;
diff --git a/src/preferences/power/PowerApp.cpp b/src/preferences/power/PowerApp.cpp
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/src/preferences/power/PowerApp.cpp
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2026, Alan Shearer, sakison@gmail.com
+ * Distributed under the terms of the MIT License.
+ */
+
+
+#include "PowerWindow.h"
+
+#include <Application.h>
+#include <Catalog.h>
+
+
+#undef B_TRANSLATION_CONTEXT
+#define B_TRANSLATION_CONTEXT "Power"
+
+
+static const char* kSignature = "application/x-vnd.Haiku-Power";
+
+
+class PowerApp : public BApplication {
+public:
+							PowerApp();
+};
+
+
+PowerApp::PowerApp()
+	:
+	BApplication(kSignature)
+{
+	PowerWindow* window = new PowerWindow();
+	window->Show();
+}
+
+
+int
+main(void)
+{
+	PowerApp app;
+	app.Run();
+	return 0;
+}
diff --git a/src/preferences/power/PowerWindow.h b/src/preferences/power/PowerWindow.h
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/src/preferences/power/PowerWindow.h
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2026, Alan Shearer, sakison@gmail.com
+ * Distributed under the terms of the MIT License.
+ */
+#ifndef POWER_WINDOW_H
+#define POWER_WINDOW_H
+
+
+#include <Window.h>
+
+
+class BCheckBox;
+
+
+// Same code power_daemon handles. Do not change it on one side only.
+enum {
+	kMsgReloadPowerSettings = 'PwRl'
+};
+
+
+class PowerWindow : public BWindow {
+public:
+							PowerWindow();
+	virtual					~PowerWindow();
+
+	virtual	void			MessageReceived(BMessage* message);
+	virtual	bool			QuitRequested();
+
+private:
+			void			_LoadSettings();
+			void			_SaveSettings();
+			void			_UpdateEnabled();
+
+			BCheckBox*		fOnBattery;
+			BCheckBox*		fKeepOnAc;
+};
+
+#endif
diff --git a/src/preferences/power/PowerWindow.cpp b/src/preferences/power/PowerWindow.cpp
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/src/preferences/power/PowerWindow.cpp
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2026, Alan Shearer, sakison@gmail.com
+ * Distributed under the terms of the MIT License.
+ */
+
+
+#include "PowerWindow.h"
+
+#include <Button.h>
+#include <Catalog.h>
+#include <CheckBox.h>
+#include <File.h>
+#include <FindDirectory.h>
+#include <LayoutBuilder.h>
+#include <Messenger.h>
+#include <Path.h>
+#include <StringView.h>
+
+
+#undef B_TRANSLATION_CONTEXT
+#define B_TRANSLATION_CONTEXT "Power"
+
+
+static const char* kSettingsFile = "power_daemon";
+static const uint32 kMsgOnBattery = 'OnBt';
+static const uint32 kMsgKeepOnAc = 'KpAc';
+
+
+PowerWindow::PowerWindow()
+	:
+	BWindow(BRect(0, 0, 100, 100), B_TRANSLATE_SYSTEM_NAME("Power"),
+		B_TITLED_WINDOW, B_NOT_ZOOMABLE | B_AUTO_UPDATE_SIZE_LIMITS
+			| B_QUIT_ON_WINDOW_CLOSE),
+	fOnBattery(NULL),
+	fKeepOnAc(NULL)
+{
+	fOnBattery = new BCheckBox("on_battery",
+		B_TRANSLATE("Power saving on battery"), new BMessage(kMsgOnBattery));
+	fKeepOnAc = new BCheckBox("keep_on_ac",
+		B_TRANSLATE("Keep power saving on AC"), new BMessage(kMsgKeepOnAc));
+
+	BStringView* note = new BStringView("note",
+		B_TRANSLATE("Used when the machine switches between battery and AC."));
+
+	BLayoutBuilder::Group<>(this, B_VERTICAL, B_USE_SMALL_SPACING)
+		.SetInsets(B_USE_WINDOW_SPACING)
+		.Add(fOnBattery)
+		.Add(fKeepOnAc)
+		.Add(note)
+		.End();
+
+	_LoadSettings();
+	_UpdateEnabled();
+	CenterOnScreen();
+}
+
+
+PowerWindow::~PowerWindow()
+{
+}
+
+
+void
+PowerWindow::MessageReceived(BMessage* message)
+{
+	switch (message->what) {
+		case kMsgOnBattery:
+		case kMsgKeepOnAc:
+			_UpdateEnabled();
+			_SaveSettings();
+			break;
+
+		default:
+			BWindow::MessageReceived(message);
+			break;
+	}
+}
+
+
+bool
+PowerWindow::QuitRequested()
+{
+	_SaveSettings();
+	return true;
+}
+
+
+void
+PowerWindow::_LoadSettings()
+{
+	bool onBattery = false;
+	bool keepOnAc = false;
+
+	BPath path;
+	if (find_directory(B_SYSTEM_SETTINGS_DIRECTORY, &path) == B_OK) {
+		path.Append(kSettingsFile);
+		BFile file(path.Path(), B_READ_ONLY);
+		BMessage settings;
+		if (file.InitCheck() == B_OK && settings.Unflatten(&file) == B_OK) {
+			settings.FindBool("on_battery", &onBattery);
+			settings.FindBool("keep_on_ac", &keepOnAc);
+		}
+	}
+
+	fOnBattery->SetValue(onBattery ? B_CONTROL_ON : B_CONTROL_OFF);
+	fKeepOnAc->SetValue(keepOnAc ? B_CONTROL_ON : B_CONTROL_OFF);
+}
+
+
+void
+PowerWindow::_SaveSettings()
+{
+	BMessage settings;
+	settings.AddBool("on_battery", fOnBattery->Value() == B_CONTROL_ON);
+	settings.AddBool("keep_on_ac", fKeepOnAc->Value() == B_CONTROL_ON);
+
+	BPath path;
+	if (find_directory(B_SYSTEM_SETTINGS_DIRECTORY, &path) != B_OK)
+		return;
+	path.Append(kSettingsFile);
+
+	BFile file(path.Path(), B_WRITE_ONLY | B_CREATE_FILE | B_ERASE_FILE);
+	if (file.InitCheck() != B_OK)
+		return;
+	if (settings.Flatten(&file) != B_OK)
+		return;
+
+	BMessenger daemon("application/x-vnd.Haiku-powermanagement");
+	if (daemon.IsValid())
+		daemon.SendMessage(kMsgReloadPowerSettings);
+}
+
+
+void
+PowerWindow::_UpdateEnabled()
+{
+	fKeepOnAc->SetEnabled(fOnBattery->Value() == B_CONTROL_ON);
+}
